src/v/cloud_storage/recursive_directory_walker.cc

```diff
@@ -107,28 +104,22 @@ ss::future<walk_result> recursive_directory_walker::walk(
     walk_accumulator state(start_dir, tracker, std::move(collect_filter));
 
     fragmented_vector<ss::sstring> empty_dirs;
-
-    while (!state.empty()) {
-        auto target = state.pop();
-        vassert(
-          std::string_view(target).starts_with(start_dir),
-          "Looking at directory {}, which is outside of initial dir "
-          "{}.",
-          target,
-          start_dir);
-
+    auto process_directory = [&start_dir, &state, &empty_dirs](
+                               ss::sstring const& target) -> ss::future<> {
```

Comment:
I think we're trying to avoid coroutine lambdas. This could be a method of the directory walker class.